packages/datadog-instrumentations/src/kafkajs.js

```diff
@@ -0,0 +1,114 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const producerStartCh = channel('apm:kafkajs:produce:start')
+const producerMessageCh = channel(`apm:kafkajs:produce:message`)
+const producerFinishCh = channel('apm:kafkajs:produce:finish')
+const producerErrorCh = channel('apm:kafkajs:produce:error')
+
+const consumerStartCh = channel('apm:kafkajs:consume:start')
+const consumerFinishCh = channel('apm:kafkajs:consume:finish')
+const consumerErrorCh = channel('apm:kafkajs:consume:error')
+
+addHook({ name: 'kafkajs', versions: ['>=1.4'] }, (obj) => {
+  const Kafka = obj.Kafka
+  shimmer.wrap(Kafka.prototype, 'producer', createProducer => function () {
+    if (!producerStartCh.hasSubscribers) {
+      return createProducer.apply(this, arguments)
+    }
+    const producer = createProducer.apply(this, arguments)
+    const send = producer.send
+
+    const outerAsyncResource = new AsyncResource('bound-anonymous-fn')
+    const innerAsyncResource = new AsyncResource('bound-anonymous-fn')
+
+    producer.send = function () {
+      return innerAsyncResource.runInAsyncScope(() => {
+        producerStartCh.publish(undefined)
+        try {
+          const lastArgId = arguments.length - 1
+          const cb = arguments[lastArgId]
+          if (typeof cb === 'function') {
+            const scopeBoundCb = outerAsyncResource.bind(cb)
+            arguments[lastArgId] = innerAsyncResource.bind(function (err) {
+              if (err) {
+                producerErrorCh.publish(err)
+              }
+              producerFinishCh.publish(undefined)
+              return scopeBoundCb.apply(this, arguments)
+            })
+          }
+          const { topic, messages = [] } = arguments[0]
+          for (const message of messages) {
+            message.headers = message.headers || {}
+          }
+          producerMessageCh.publish({ topic, messages })
+
+          const result = send.apply(this, arguments)
+
+          return result
+        } catch (e) {
+          producerErrorCh.publish(e)
+          producerFinishCh.publish(undefined)
+          throw e
+        } finally {
+          producerFinishCh.publish(undefined)
```

Comment:
Why is `finish` called here if it's meant to signal that the message has been sent? Is it for when there are no callbacks? Right now this would run either way.